tests/Application/Actions/Donations/UpdateHandlesLockExceptionTest.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MatchBot\Tests\Application\Actions\Donations;
+
+use Doctrine\DBAL\Exception\DriverException;
+use Doctrine\DBAL\Exception\LockWaitTimeoutException;
+use Doctrine\ORM\EntityManagerInterface;
+use GuzzleHttp\Psr7\ServerRequest;
+use MatchBot\Application\Actions\Donations\Update;
+use MatchBot\Domain\Campaign;
+use MatchBot\Domain\Charity;
+use MatchBot\Domain\Donation;
+use MatchBot\Domain\DonationRepository;
+use Prophecy\PhpUnit\ProphecyTrait;
+use Psr\Log\NullLogger;
+use Ramsey\Uuid\Uuid;
+use Slim\Psr7\Response;
+use Stripe\Service\PaymentIntentService;
+use Stripe\StripeClientInterface;
+use Symfony\Component\Serializer\Encoder\JsonEncoder;
+use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;
+use Symfony\Component\Serializer\Serializer;
+
+class UpdateHandlesLockExceptionTest extends \PHPUnit\Framework\TestCase
+{
+    use ProphecyTrait;
+
+    public function testRetryOnLockWaitTimeOut(): void
+    {
+        // arrange
+        $donationId = 'donation_id';
+
+        $donationRepositoryProphecy = $this->prophesize(DonationRepository::class);
+        $entityManagerProphecy = $this->prophesize(EntityManagerInterface::class);
+        $stripeIntentsProphecy = $this->prophesize(PaymentIntentService::class);
+        $fakeStripeClient = $this->fakeStripeClient($stripeIntentsProphecy->reveal());
+
+        $charity = new Charity();
+        $charity->setDonateLinkId('DONATE_LINK_ID');
+        $charity->setName('Charity name');
+
+        $campaign = new Campaign();
+        $campaign->setIsMatched(true);
+        $campaign->setCharity($charity);
+
+        $donation = new Donation();
+        $donation->createdNow();
+        $donation->setCampaign($campaign);
+        $donation->setPsp('stripe');
+        $donation->setCurrencyCode('GBP');
+        $donation->setAmount('1');
+        $donation->setUuid(Uuid::uuid4());
+        $donation->setDonationStatus('Pending');
+        $donation->setDonorFirstName('Donor first name');
+        $donation->setDonorLastName('Donor last name');
```

Comment:
it was a bit of a slow process finding all the setters I needed to call to stop TypeErrors. At some point when we're doing more work on MatchBot I'd like to make constructors require whatever parameters are necessary to make a usable entity, so we don't have to rely on setters.

Reply:
I think this is a useful suggestion

Reply:
Thanks. This is one of the things that Psalm would partially enforce for new code. I like to follow the principle that [Objects should be constructed in one go](https://matthiasnoback.nl/2018/07/objects-should-be-constructed-in-one-go/).
Something to look at as we continue to work on this code.